server/service/src/invoice/inbound_shipment/add_from_master_list.rs

```diff
@@ -0,0 +1,307 @@
+use crate::invoice::common::{
+    AddToShipmentFromMasterListError as ServiceError,
+    AddToShipmentFromMasterListInput as ServiceInput,
+};
+use crate::invoice::{check_invoice_exists, common::check_master_list_for_name};
+use crate::service_provider::ServiceContext;
+use crate::sync_processor::invoice::common::get_lines_for_invoice;
+use repository::EqualFilter;
+use repository::{
+    InvoiceLine, InvoiceLineFilter, InvoiceLineRepository, InvoiceLineRow,
+    InvoiceLineRowRepository, InvoiceRow, InvoiceRowStatus, InvoiceRowType, MasterListLineFilter,
+    MasterListLineRepository, RepositoryError, StorageConnection,
+};
+
+use super::generate_unallocated_invoice_lines;
+
+pub fn add_from_master_list(
+    ctx: &ServiceContext,
+    store_id: &str,
+    input: ServiceInput,
+) -> Result<Vec<InvoiceLine>, ServiceError> {
+    let invoice_lines = ctx
+        .connection
+        .transaction_sync(|connection| {
+            let invoice_row = validate(connection, store_id, &input)?;
+            let new_invoice_line_rows = generate(ctx, invoice_row, &input)?;
+
+            let invoice_line_row_repository = InvoiceLineRowRepository::new(&connection);
+
+            for invoice_line_row in new_invoice_line_rows {
+                invoice_line_row_repository.upsert_one(&invoice_line_row)?;
+            }
+
+            match InvoiceLineRepository::new(connection).query_by_filter(
+                InvoiceLineFilter::new().invoice_id(EqualFilter::equal_to(&input.shipment_id)),
+            ) {
+                Ok(lines) => Ok(lines),
+                Err(error) => Err(ServiceError::DatabaseError(error)),
+            }
+        })
+        .map_err(|error| error.to_inner_error())?;
+    Ok(invoice_lines)
+}
+
+fn validate(
+    connection: &StorageConnection,
+    store_id: &str,
+    input: &ServiceInput,
+) -> Result<InvoiceRow, ServiceError> {
+    let invoice_row = match check_invoice_exists(&input.shipment_id, connection) {
+        Ok(row) => row,
+        Err(_error) => return Err(ServiceError::ShipmentDoesNotExist),
+    };
+
+    if invoice_row.store_id != store_id {
+        return Err(ServiceError::NotThisStoreShipment);
+    }
+    if invoice_row.status == InvoiceRowStatus::Shipped
+        || invoice_row.status == InvoiceRowStatus::Delivered
+        || invoice_row.status == InvoiceRowStatus::Verified
+    {
+        return Err(ServiceError::CannotEditShipment);
+    }
+
+    if invoice_row.r#type != InvoiceRowType::InboundShipment {
+        return Err(ServiceError::NotAnInboundShipment);
+    }
+
+    check_master_list_for_name(connection, &invoice_row.name_id, &input.master_list_id)?
```

Comment:
hmm, we've never had inbound shipment master list logic before, so the question is, should it be for store or for name ?

I am leaning towards `store` since that's what requisitions do (i.e. for request it's for store, for response it's for. name). What do you think ? Should we ask AV ?

Reply:
store sounds sensible to me.. I'll update